src/nav2_navigation/bt_navigator/src/BTNavigator.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <exception>
+#include <chrono>
+#include "bt_navigator/BTNavigator.hpp"
+
+using nav2_tasks::TaskStatus;
+
+namespace bt_navigator
+{
+
+BTNavigator::BTNavigator(const std::string & name)
+: nav2_tasks::NavigateToPoseTaskServer(name)
+{
+  RCLCPP_INFO(get_logger(), "BTNavigator::BTNavigator");
+  planner_ = std::make_unique<nav2_tasks::ComputePathToPoseTaskClient>("AStarPlanner", this);
+  controller_ = std::make_unique<nav2_tasks::FollowPathTaskClient>("DwaController", this);
+}
+
+BTNavigator::~BTNavigator()
+{
+  RCLCPP_INFO(get_logger(), "BTNavigator::~BTNavigator");
+}
+
+TaskStatus
+BTNavigator::executeAsync(const nav2_tasks::NavigateToPoseCommand::SharedPtr command)
```

Comment:
This function and the executeAsync in simpleNavigator are very long and could be easily broken up into smaller logical chunks. I think if you do that, you could probably eliminate the need for that goto as well.

Reply:
Behavior trees handle this nicely for the BTNavigator. I'll see about making a function for the SimpleNavigator.